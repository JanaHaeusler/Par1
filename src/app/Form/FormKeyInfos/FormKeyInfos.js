import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import Button from '../../Button'
import { CancelIconDark, CheckIconLight } from '../../Icons/Icons'

FormKeyInfos.propTypes = {
  formInputs: PropTypes.object.isRequired,
  isSaveButtonShown: PropTypes.bool.isRequired,
  updateDirtyInputs: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  showErrorMessage: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired,
}

export default function FormKeyInfos({
  formInputs,
  isSaveButtonShown,
  updateDirtyInputs,
  handleChange,
  showErrorMessage,
  handleSubmit,
  handleCancel,
}) {
  return (
    <Form noValidate onSubmit={handleSubmit} data-testid="form">
      <h3>Info</h3>
      <KeyInfos>
        <label>
          Location
          <input
            type="text"
            name="location"
            id="location"
            placeholder="Type location ..."
            value={formInputs.location}
            onChange={(event) =>
              handleChange(event.target.name, event.target.value)
            }
            onBlur={() => updateDirtyInputs('location')}
          />
        </label>
        <Note>{showErrorMessage('location')}</Note>

        <label>
          Date
          <input
            type="date"
            name="date"
            id="date"
            value={formInputs.date}
            onChange={(event) =>
              handleChange(event.target.name, event.target.value)
            }
            onBlur={() => updateDirtyInputs('date')}
          />
        </label>
        <Note>{showErrorMessage('date')}</Note>

        <label>
          Player(s)
          <input
            type="text"
            name="players"
            id="players"
            placeholder="John, Jane"
            value={formInputs.players}
            onChange={(event) =>
              handleChange(event.target.name, event.target.value)
            }
            onBlur={() => updateDirtyInputs('players')}
          />
        </label>
        <Note>{showErrorMessage('players')}</Note>

        <label>
          Winner(s)
          <input
            type="text"
            name="winner"
            id="winner"
            placeholder="Jane"
            value={formInputs.winner}
            onChange={(event) =>
              handleChange(event.target.name, event.target.value)
            }
            onBlur={() => updateDirtyInputs('winner')}
          />
        </label>
        <Note>{showErrorMessage('winner')}</Note>

        <label>
          Total Shots Winner(s)
          <input
            type="number"
            name="shots"
            id="shots"
            placeholder="38"
            value={formInputs.shots}
            onChange={(event) =>
              handleChange(event.target.name, event.target.value)
            }
            onBlur={() => updateDirtyInputs('shots')}
          />
        </label>
        <Note>{showErrorMessage('shots')}</Note>
      </KeyInfos>
      <ButtonWrapper>
        <Button
          type="button"
          onClick={handleCancel}
          iconComponent={<CancelIconDark />}
          text="Cancel"
          data-testid="button-cancel"
        />
        <Button
          main
          disabled={!isSaveButtonShown}
          iconComponent={<CheckIconLight />}
          text="Save"
          data-testid="button-save"
        />
      </ButtonWrapper>
      <Note>
        *Please do not clear your browsers cache, in order to permanently save
        your game details
      </Note>
    </Form>
  )
}

const Form = styled.form`
  margin: 10px 20px;
  padding: 10px;
  display: grid;
  place-items: center;
  border-radius: 25px;
  background-color: var(--light);
  font-size: 1rem;

  input::-webkit-outer-spin-button,
  input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  input[type='number'] {
    -moz-appearance: textfield;
  }

  Button {
    margin-top: 15px;
    margin-bottom: 5px;
  }
`
const KeyInfos = styled.fieldset`
  margin: 0;
  padding: 0;
  border: none;
  display: grid;
  width: 100%;

  label {
    margin-top: 10px;
    margin-bottom: 5px;
  }

  input {
    display: block;
    margin-top: 3px;
    padding: 5px;
    width: 100%;
    border-style: none;
    border-bottom: 1px solid var(--secondary-medium);
    color: var(--secondary-medium);
    font-family: 'Montserrat', sans-serif;
  }
`
const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
  margin-bottom: 10px;
  width: 100%;
`
const Note = styled.span`
  margin-top: 5px;
  font-size: 0.7rem;

  :last-child {
    margin-bottom: 10px;
  }
`
