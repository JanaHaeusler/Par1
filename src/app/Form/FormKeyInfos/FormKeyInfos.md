Form Key Infos

```jsx
<FormKeyInfos
  formInputs={{
    location: '',
    date: '',
    players: '',
    winner: '',
    shots: '',
  }}
  isSaveButtonShown={false}
  updateDirtyInputs={() => {}}
  handleChange={() => {}}
  showErrorMessage={() => {}}
  handleSubmit={() => {}}
  handleCancel={() => {}}
  changeDisabledModus={() => {}}
/>
```
