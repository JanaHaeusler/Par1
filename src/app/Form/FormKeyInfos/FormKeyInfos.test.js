import { fireEvent, render } from '@testing-library/react'
import user from '@testing-library/user-event'
import FormKeyInfos from './FormKeyInfos'

const testProps = {
  formInputs: {
    location: '',
    date: '',
    players: '',
    winner: '',
    shots: '',
  },
  isSaveButtonShown: false,
  updateDirtyInputs: jest.fn(),
  handleChange: jest.fn(),
  showErrorMessage: jest.fn(),
  handleSubmit: jest.fn(),
  handleCancel: jest.fn(),
  setDisabledState: jest.fn(),
}

describe('FormKeyInfos', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => null),
        setItem: jest.fn(() => null),
      },
      writable: true,
    })
  })

  it('renders snapshot correctly', () => {
    const { container } = render(<FormKeyInfos {...testProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders correctly', () => {
    const { getByLabelText, getByTestId } = render(
      <FormKeyInfos {...testProps} />
    )
    expect(getByLabelText('Location')).toBeInTheDocument()
    expect(getByLabelText('Date')).toBeInTheDocument()
    expect(getByLabelText('Player(s)')).toBeInTheDocument()
    expect(getByLabelText('Winner(s)')).toBeInTheDocument()
    expect(getByLabelText('Total Shots Winner(s)')).toBeInTheDocument()

    expect(getByLabelText('Location')).toHaveValue('')
    expect(getByLabelText('Date')).toHaveValue('')
    expect(getByLabelText('Player(s)')).toHaveValue('')
    expect(getByLabelText('Winner(s)')).toHaveValue('')
    expect(getByLabelText('Total Shots Winner(s)')).toHaveValue(null)

    expect(getByTestId('button-save')).toBeDisabled()
    expect(getByTestId('button-cancel')).toBeInTheDocument()
  })

  it('has enabled save button', () => {
    const props = { ...testProps, isSaveButtonShown: true }
    const { getByTestId } = render(<FormKeyInfos {...props} />)
    expect(getByTestId('button-save')).toBeEnabled()
  })

  it('calls updateDirtyInputsMock on blur event', () => {
    const updateDirtyInputsMock = jest.fn()
    const props = { ...testProps, updateDirtyInputs: updateDirtyInputsMock }
    const { getByLabelText } = render(<FormKeyInfos {...props} />)
    const InputLocation = getByLabelText('Location')
    fireEvent.blur(InputLocation)
    expect(updateDirtyInputsMock).toHaveBeenCalled()
  })

  it('calls handleChangeMock on change event', () => {
    const handleChangeMock = jest.fn()
    const props = { ...testProps, handleChange: handleChangeMock }
    const { getByLabelText } = render(<FormKeyInfos {...props} />)
    const InputPlayers = getByLabelText('Player(s)')
    fireEvent.change(InputPlayers, { target: { value: 'test' } })
    expect(handleChangeMock).toHaveBeenCalledTimes(1)
    expect(handleChangeMock).toHaveBeenCalledWith('players', 'test')
  })

  it('calls showErrorMessageMock', () => {
    const showErrorMessageMock = jest.fn()
    const props = { ...testProps, showErrorMessage: showErrorMessageMock }
    render(<FormKeyInfos {...props} />)
    expect(showErrorMessageMock).toHaveBeenCalledTimes(5)
  })

  it('calls handleSubmitMock on submitting the form', () => {
    const handleSubmitMock = jest.fn()
    const props = {
      ...testProps,
      isSaveButtonShown: true,
      handleSubmit: handleSubmitMock,
    }
    const { getByTestId } = render(<FormKeyInfos {...props} />)
    const Form = getByTestId('form')
    fireEvent.submit(Form)
    expect(handleSubmitMock).toHaveBeenCalled()
  })

  it('calls handleCancelMock on clicking cancel button', () => {
    const handleCancelMock = jest.fn()
    const props = {
      ...testProps,
      handleCancel: handleCancelMock,
    }
    const { getByTestId } = render(<FormKeyInfos {...props} />)
    user.click(getByTestId('button-cancel'))
    expect(handleCancelMock).toHaveBeenCalled()
  })

  it('calls setDisabledStateMock when rendered', () => {
    const setDisabledStateMock = jest.fn()
    const props = {
      ...testProps,
      setDisabledState: setDisabledStateMock,
    }
    render(<FormKeyInfos {...props} />)
    expect(setDisabledStateMock).toHaveBeenCalledTimes(1)
  })
})
