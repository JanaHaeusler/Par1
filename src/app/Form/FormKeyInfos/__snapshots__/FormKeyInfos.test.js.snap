// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormKeyInfos renders snapshot correctly 1`] = `
.c4 {
  padding: 7px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  width: 85px;
  height: 34px;
  border-style: none;
  border-radius: 5px;
  box-shadow: 2px 2px 8px var(--secondary-medium-transparent);
  background-color: var(--light);
  color: var(--primary);
  font-family: 'Raleway',sans-serif;
  font-size: 1rem;
  font-weight: 550;
}

.c4:disabled {
  opacity: 50%;
}

.c6 {
  padding: 7px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  width: 85px;
  height: 34px;
  border-style: none;
  border-radius: 5px;
  box-shadow: 2px 2px 8px var(--secondary-medium-transparent);
  background-color: var(--primary);
  color: var(--light);
  font-family: 'Raleway',sans-serif;
  font-size: 1rem;
  font-weight: 550;
}

.c6:disabled {
  opacity: 50%;
}

.c5 {
  margin-right: 3px;
  stroke: var(--primary);
}

.c7 {
  margin-right: 3px;
  width: 25%;
  stroke: var(--light);
}

.c0 {
  margin: 10px 20px;
  padding: 10px;
  display: grid;
  place-items: center;
  border-radius: 25px;
  background-color: var(--light);
  font-size: 1rem;
}

.c0 input::-webkit-outer-spin-button,
.c0 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c0 input[type='number'] {
  -moz-appearance: textfield;
}

.c0 Button {
  margin-top: 15px;
  margin-bottom: 5px;
}

.c1 {
  margin: 0;
  padding: 0;
  border: none;
  display: grid;
  width: 100%;
}

.c1 label {
  margin-top: 10px;
  margin-bottom: 5px;
}

.c1 input {
  display: block;
  margin-top: 3px;
  padding: 5px;
  width: 100%;
  border-style: none;
  border-bottom: 1px solid var(--secondary-medium);
  color: var(--secondary-medium);
  font-family: 'Montserrat',sans-serif;
}

.c1 input::-webkit-input-placeholder {
  font-style: italic;
  opacity: 0.8;
}

.c1 input::-moz-placeholder {
  font-style: italic;
  opacity: 0.8;
}

.c1 input:-ms-input-placeholder {
  font-style: italic;
  opacity: 0.8;
}

.c1 input::placeholder {
  font-style: italic;
  opacity: 0.8;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  margin-bottom: 10px;
  width: 100%;
}

.c2 {
  margin-top: 5px;
  font-size: 0.7rem;
}

.c2:last-child {
  margin-bottom: 10px;
}

<form
  class="c0"
  data-testid="form"
  novalidate=""
>
  <h3>
    Info
  </h3>
  <fieldset
    class="c1"
  >
    <label>
      Location
      <input
        id="location"
        name="location"
        placeholder="Name of location ..."
        type="text"
        value=""
      />
    </label>
    <span
      class="c2"
    />
    <label>
      Date
      <input
        id="date"
        name="date"
        type="date"
        value=""
      />
    </label>
    <span
      class="c2"
    />
    <label>
      Player(s)
      <input
        id="players"
        name="players"
        placeholder="John, Jane, ..."
        type="text"
        value=""
      />
    </label>
    <span
      class="c2"
    />
    <label>
      Winner(s)
      <input
        id="winner"
        name="winner"
        placeholder="Jane"
        type="text"
        value=""
      />
    </label>
    <span
      class="c2"
    />
    <label>
      Total Shots Winner(s)
      <input
        id="shots"
        name="shots"
        placeholder="38"
        type="number"
        value=""
      />
    </label>
    <span
      class="c2"
    />
  </fieldset>
  <div
    class="c3"
  >
    <button
      class="c4"
      data-testid="button-cancel"
      type="button"
    >
      <svg
        class="c5"
      >
        cross.svg
      </svg>
      Cancel
    </button>
    <button
      class="c6"
      data-testid="button-save"
      disabled=""
    >
      <svg
        class="c7"
      >
        check.svg
      </svg>
      Save
    </button>
  </div>
  <span
    class="c2"
  >
    *Please do not clear your browsers cache, in order to permanently save your game details
  </span>
</form>
`;
